src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationPerformanceMonitor.cs

```diff
@@ -2,89 +2,156 @@
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
 using System;
+using System.Collections.Generic;
 using System.Diagnostics;
 using System.IO;
+using System.Linq;
 using UnityEngine;
 
 namespace Microsoft.MixedReality.SpectatorView
 {
-    internal enum StateSynchronizationPerformanceFeature : byte
+    public class StateSynchronizationPerformanceMonitor : Singleton<StateSynchronizationPerformanceMonitor>
     {
-        GameObjectComponentCheck = 0x0,
-        MaterialPropertyUpdate = 0x1,
-        MaterialPropertyBlockUpdate = 0x2,
-        Count = 0x3
-    }
-
-    internal class StateSynchronizationPerformanceMonitor : Singleton<StateSynchronizationPerformanceMonitor>
-    {
-        private const int PeriodsToAverageOver = 5;
-        private const int FeatureCount = (int)StateSynchronizationPerformanceFeature.Count;
-        private Stopwatch[] stopwatches;
-        private float[][] previousSpentTimes;
-        private float[][] previousActualTimes;
-        private int currentPeriod = 0;
+        private Dictionary<string, Dictionary<string, Stopwatch>> eventStopWatches = new Dictionary<string, Dictionary<string, Stopwatch>>();
+        private Dictionary<string, Dictionary<string, int>> eventCounts = new Dictionary<string, Dictionary<string, int>>();

```

Comment:
For both readability and performance, would prefer if there was a struct TimerKey with two fields (ComponentName and EventName) that can be used as the key for both of these dictionaries. This would avoid the extra tier of allocations for all of the internal dictionaries, and would make it easier to understand lookups when reading the code.